src/AMSlib/AMS.cpp

```diff
@@ -5,22 +5,392 @@
+    }
+
+    if (!value.contains("threshold")) {
+      THROW(std::runtime_error,
+            "Model must define threshold value (threshold < 0 always "
+            "performs original code, threshold=1e30 always use the "
+            "model)");
+    }
+    threshold = value["threshold"].get<float>();
+    parseUQPaths(uqPolicy, value);
+    DBLabel = parseDBLabel(value);
+  }
+
+
+  AMSAbstractModel(AMSUQPolicy uq_policy,
+                   const char *surrogate_path,
+                   const char *uq_path,
+                   const char *db_label,
+                   double threshold,
+                   int num_clusters)
+  {
+    if (db_label == nullptr)
+      FATAL(AMS, "registering model without a database identifier\n");
+
+    DBLabel = std::string(db_label);
+
+    if (!BaseUQ::isUQPolicy(uq_policy)) {
+      FATAL(AMS, "Invalid UQ policy %d", uq_policy)
+    }
+
+    uqPolicy = uq_policy;
+
+    if (surrogate_path != nullptr) SPath = std::string(surrogate_path);
+
+    if (uq_path != nullptr) UQPath = std::string(uq_path);
+
+    this->threshold = threshold;
+    nClusters = num_clusters;
+    DBG(AMS,
+        "Registered Model %s %g",
+        BaseUQ::UQPolicyToStr(uqPolicy).c_str(),
+        threshold);
+  }
+
+
+  void dump()
+  {
+    if (!SPath.empty()) DBG(AMS, "Surrogate Model Path: %s", SPath.c_str());
+    if (!UQPath.empty()) DBG(AMS, "UQ-Model: %s", UQPath.c_str());
+    DBG(AMS,
+        "db-Label: %s threshold %f UQ-Policy: %u nClusters: %d",
+        DBLabel.c_str(),
+        threshold,
+        uqPolicy,
+        nClusters);
+  }
+};
+
+
+class AMSWrap
+{
+  using json = nlohmann::json;
+
+public:
   std::vector<std::pair<AMSDType, void *>> executors;
+  std::vector<AMSAbstractModel> registered_models;
+  std::unordered_map<std::string, int> ams_candidate_models;
+  AMSDBType dbType = AMSDBType::AMS_NONE;
+
+private:
+  void dumpEnv()
+  {
+    for (auto &KV : ams_candidate_models) {
+      DBG(AMS, "\n")
+      DBG(AMS,
+          "\t\t\t Model: %s With AMSAbstractID: %d",
+          KV.first.c_str(),
+          KV.second);
+      if (KV.second >= ams_candidate_models.size()) {
+        FATAL(AMS,
+              "Candidate model mapped to AMSAbstractID that does not exist "
+              "(%d)",
+              KV.second);
+      }
+      auto &abstract_model = registered_models[KV.second];
+      abstract_model.dump();
+    }
+  }
+
+  void parseDomainModels(
+      json &jRoot,
+      std::unordered_map<std::string, std::string> &domain_map)
+  {
+    if (!jRoot.contains("domain_models")) return;
+
+    auto domain_models = jRoot["domain_models"];
+    for (auto &field : domain_models.items()) {
+      auto &name = field.key();
+      auto val = field.value().get<std::string>();
+      domain_map.emplace(val, name);
+    }
+    return;
+  }
+
+  void parseCandidateAMSModels(
+      json &jRoot,
+      std::unordered_map<std::string, std::string> ml_domain_mapping)
+  {
+    if (jRoot.contains("ml_models")) {
```

Comment:
could simplify nesting if changing conditional to
```
if (!jRoot.contains("ml_models"))
  return
```
may apply to other methods later

Reply:
Fixed.